TESTING.md

```diff
@@ -0,0 +1,123 @@
+# Testing guidelines
+
+Following guide will help you decide what to test and how to do go about it
+when changing code in this repository.
+
+## Testing levels
+
+In KIC, we use several levels of testing:
+
+- [unit tests](#unit-tests)
+- [integration tests](#integration-tests)
+- [end to end (E2E) tests](#end-to-end-e2e-tests)
+
+### Unit tests
+
+Unit tests, verify the functionality of one or more `func` or `struct`s
+in an isolated environment.
+
+These tests are independent of the environment they are run in and require no
+setup of teardown code.
+
+> **NOTE**: Some of these tests do require [`envtest`][envtest] setup to be present e.g.
+> [this one][test_GetAdminAPIsForService] or [this one][test_HTTPRouteReconcilerProperlyReactsToReferenceGrant].
+> This setup is handled automatically via `setup-envtest` target which is a dependency
+> of mentioned above unit test targets.
+>
+> As part of [#4099][issue4099] this organization might change and test that do
+> require this setup might end up in a separate place.
+
+These tests can either be written in a way that do not require any mocks or fakes,
+or using e.g. kubernetes related fakes like for instance
+[`TestAddressesFromEndpointSlice`][test_AddressesFromEndpointSlice] does.
+
+#### How to run
+
+Unit tests can be either run via appropriate `go test ...` invocation or using any of
+the provided Makefile targets:
+
+- `test.unit` run unit tests with standard verbose output on stderr
+- `test.unit.pretty` run unit tests with opinionated output format, one line per package
+
+Unit tests are typically located in the same package as the code they test
+and they do not use any [build tags][go_build_tag].
+
+> **NOTE**: Some of these tests that do require [`envtest`][envtest] setup do
+> use a special [build tags][go_build_tag]: `envtest`. This is in place so that
+> simply running `go test ...` without any special build tags would pass without
+> requiring any manual setup from the developer.
```

Comment:
```suggestion
> **NOTE**: Some of these tests that do require [`envtest`][envtest] setup and
> use a special [build tags][go_build_tag]: `envtest`. This is in place so that
> simply running `go test ...` without any special build tags would pass without
> requiring any manual setup from the developer.
```